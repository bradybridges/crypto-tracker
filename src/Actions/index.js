export const updateCryptos = (cryptos) => ({
  type: 'UPDATE_CRYPTOS',
  cryptos,
});

export const updateError = (error) => ({
  type: 'UPDATE_ERROR',
  error,
});

export const updateTimeFrame = (timeFrame) => ({
  type: 'UPDATE_TIME_FRAME',
  timeFrame,
});

export const updateTrackedCoins = (symbols) => ({
  type: 'UPDATE_TRACKED_COINS',
  symbols,
});

export const updatePortfolio = (portfolio) => ({
  type: 'UPDATE_PORTFOLIO',
  portfolio,
});
