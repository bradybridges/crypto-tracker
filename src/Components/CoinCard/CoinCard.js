import React from 'react';
import './CoinCard.scss';

const CoinCard = ({name, rank, logo, price}) => {
  return (
    <section>
      <h1>{name}</h1>
      <p>Rank: {rank}</p>
      <img src={logo} alt='coin logo' />
      <p>${Number(price).toFixed(2)}</p>
      <button>Track Coin</button>
    </section>
  );
}

export default CoinCard;

// 1d: {price_change: "-10.35376944", price_change_pct: "-0.00113751", volume: "16486354356.23", volume_change: "-1060914704.34", volume_change_pct: "-0.06", …}
// 1h: {price_change: "-37.12686146", price_change_pct: "-0.00406694", volume: "542724905.47", volume_change: "-20801737.02", volume_change_pct: "-0.04", …}
// 30d: {price_change: "778.71412728", price_change_pct: "0.09367306", volume: "396040290759.14", volume_change: "54556791262.92", volume_change_pct: "0.16", …}
// 365d: {price_change: "2661.32712864", price_change_pct: "0.41386049", volume: "3241626349070.16", volume_change: "1533942003592.29", volume_change_pct: "0.90", …}
// circulating_supply: "18023112"
// currency: "BTC"
// high: "19408.82122873"
// high_timestamp: "2017-12-16"
// id: "BTC"
// logo_url: "https://s3.us-east-2.amazonaws.com/nomics-api/static/images/currencies/btc.svg"
// market_cap: "163862899935.33"
// max_supply: "21000000"
// name: "Bitcoin"
// price: "9091.82054327"
// price_date: "2019-11-01"
// rank: "1"
// symbol: "BTC"