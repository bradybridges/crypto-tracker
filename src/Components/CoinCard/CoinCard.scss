table {
  width: 50%;
  justify-self: center;
  border-collapse: collapse;
  margin: 0 auto;
}

#coin-card-section {
  width: 90%;
  text-align: center;
}