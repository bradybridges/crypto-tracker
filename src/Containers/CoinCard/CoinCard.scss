table {
  width: 50%;
  justify-self: center;
  border-collapse: collapse;
  margin: 10px auto;
}

#coin-card-section {
  width: 90%;
  text-align: center;
  background-color: #383838e0;
  border-radius: 4px;
}

.tracking-btn {
  font-size: 1em;
  background-color: black;
  color: white;
  height: 50px;
  border-radius: 4px;
  margin-bottom: 10px;
}