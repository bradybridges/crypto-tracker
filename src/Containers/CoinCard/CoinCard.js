import React, { Component } from 'react';
import './CoinCard.scss';
import { connect } from 'react-redux';
import { updateTrackedCoins, updateCryptos, updatePortfolio } from '../../Actions/index';

export class CoinCard extends Component {

  trackCoin = (e) => {
    e.preventDefault();
    const updatedSymbols = [...this.props.trackedCoins, this.props.coin.id];
    this.props.updateTrackedCoins(updatedSymbols);
    this.props.getCryptos(updatedSymbols);
    localStorage.setItem('trackedCoins', JSON.stringify(updatedSymbols));
  }

  stopTrackingCoin = (e) => {
    e.preventDefault();
    const index = this.props.cryptos.findIndex(coin => coin.name === this.props.name);
    let updatedCryptos = this.props.cryptos.map(crypto => crypto);
    updatedCryptos.splice(index, 1);
    this.props.updateCryptos(updatedCryptos);
    const updatedSymbols = updatedCryptos.map( coin => coin.id);
    this.props.updateTrackedCoins(updatedSymbols);
    const updatedPortfolio = this.props.portfolio.filter(coin => coin.name !== this.props.name);
    this.props.updatePortfolio(updatedPortfolio);
    localStorage.setItem('trackedCoins', JSON.stringify(updatedSymbols));
    localStorage.setItem('portfolio', JSON.stringify(updatedPortfolio));
  }

  render() {
    const {name, logo, price, circulatingSupply} = this.props;
    const isCoinTracked = this.props.cryptos.find(coin => coin.name === name);
    return (
      <section id='coin-card-section'>
        <h1>{name}</h1>
        <img src={logo} alt='coin logo' />
        <table>
          <tbody>
            <tr>
              <td>Price</td>
              <td>${Number(price).toFixed(2)}</td>
            </tr>
            <tr>
              <td>Circulation</td>
              <td>{circulatingSupply}</td>
            </tr>
          </tbody>
        </table>
        {!isCoinTracked && <button className='tracking-btn' onClick={this.trackCoin}>Track Coin</button>}
        {isCoinTracked && <button className='tracking-btn' onClick={this.stopTrackingCoin}>Stop Tracking Coin</button>}
      </section>
    );
  }
}

export const mapDispatchToProps = dispatch => ({
  updateTrackedCoins: symbols => dispatch( updateTrackedCoins(symbols) ),
  updateCryptos: coins => dispatch( updateCryptos(coins) ),
  updatePortfolio: portfolio => dispatch( updatePortfolio(portfolio) ),
});

export const mapStateToProps = state => ({
  trackedCoins: state.trackedCoins,
  cryptos: state.cryptos,
  portfolio: state.portfolio,
});

export default connect(mapStateToProps, mapDispatchToProps)(CoinCard);