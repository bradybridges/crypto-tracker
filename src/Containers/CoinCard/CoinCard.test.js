import React from 'react';
import { shallow } from 'enzyme';
import { CoinCard, mapStateToProps, mapDispatchToProps } from './CoinCard';
import * as actions from '../../Actions/index';

describe('CoinCard', () => {
  const mockCryptos = [
    {
      currecy: 'BTC',
      id: 'BTC',
      price: 9341.6987,
      price_date: '2019-11-02',
      symbol: 'BTC',
      circulating_supply: '18025852',
      max_supply: '21000000',
      name: 'Bitcoin',
      logo_url: 'https://s3.us-east-2.amazonaws.com/nomics-api/static/images/currencies/btc.svg',
      market_cap: '1243423423.05',
      rank: '1',
      high: '19480.90',
      high_timestamp: '2017-12-16',
    },
    {
      currecy: 'LTC',
      id: 'LTC',
      price: 75.55,
      price_date: '2019-11-02',
      symbol: 'LTC',
      circulating_supply: '18000000',
      max_supply: '20000000',
      name: 'Litecoin',
      logo_url: 'https://s3.us-east-2.amazonaws.com/nomics-api/static/images/currencies/ltc.svg',
      market_cap: '1003423423.05',
      rank: '3',
      high: '420.90',
      high_timestamp: '2017-12-10',
    },
  ];
  const mockTrackedCoins = ['BTC', 'LTC'];
  const mockUpdateTrackedCoins = jest.fn();
  const mockUpdateCryptos = jest.fn();
  const mockPortfolio = [{ name: 'Bitcoin', qty: 10}, {name: 'Litecoin', qty: 5 }];
  const wrapper = shallow(<CoinCard
    coin={{ id: 'XMR' }}
    name="Monero"
    logo="https://s3.us-east-2.amazonaws.com/nomics-api/static/images/currencies/xmr.svg"
    price="1003.432342"
    circulatingSupply="123123413"
    cryptos={mockCryptos}
    trackedCoins={mockTrackedCoins}
    updateTrackedCoins={mockUpdateTrackedCoins}
    getCryptos={mockUpdateCryptos}
    updateCryptos={jest.fn()}
    updatePortfolio={jest.fn()}
    portfolio={mockPortfolio}
  />);

  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('trackDoin should call updateTrackedCoins & getCryptos', () => {
    const mockEvent = {
      preventDefault: jest.fn(),
    };
    const expected = ['BTC','LTC','XMR']
    wrapper.instance().trackCoin(mockEvent);
    expect(wrapper.instance().props.updateTrackedCoins).toHaveBeenCalledWith(expected);
    expect(wrapper.instance().props.getCryptos).toHaveBeenCalledWith(expected);
  });

  it('stopTrackingCoin should remove coin from cryptos and trackedcoins', () => {
    const mockPortfolio = [{ name: 'Bitcoin', qty: 10 }, { name: 'Litecoin', qty: 5 }];
    const wrapper2 = shallow(<CoinCard
      coin={{ id: 'BTC' }}
      name="Bitcoin"
      logo="https://s3.us-east-2.amazonaws.com/nomics-api/static/images/currencies/btc.svg"
      price="1003.432342"
      circulatingSupply="123123413"
      cryptos={mockCryptos}
      trackedCoins={mockTrackedCoins}
      updateTrackedCoins={mockUpdateTrackedCoins}
      updateCryptos={mockUpdateCryptos}
      portfolio={mockPortfolio}
      updatePortfolio={jest.fn()}
    />);
    const expectedCryptos = [
      {
        circulating_supply: '18000000',
        currecy: 'LTC',
        high: '420.90',
        high_timestamp: '2017-12-10',
        id: 'LTC',
        logo_url: 'https://s3.us-east-2.amazonaws.com/nomics-api/static/images/currencies/ltc.svg',
        market_cap: '1003423423.05',
        max_supply: '20000000',
        name: 'Litecoin',
        price: 75.55,
        price_date: '2019-11-02',
        rank: '3',
        symbol: 'LTC',
      },
    ];
    wrapper2.instance().props.updateCryptos.mockClear();
    wrapper2.instance().props.updateTrackedCoins.mockClear();
    wrapper2.instance().stopTrackingCoin({ preventDefault: jest.fn() });
    expect(wrapper2.instance().props.updateCryptos).toHaveBeenCalledWith(expectedCryptos);
    expect(wrapper2.instance().props.updateTrackedCoins).toHaveBeenCalledWith(['LTC']);
  });
}); 

describe('mapStateToProps', () => {
  it('should return trackedCoins and cryptos', () => {
    const mockCryptos = [
      {
        currecy: 'BTC',
        id: 'BTC',
        price: 9341.6987,
        price_date: '2019-11-02',
        symbol: 'BTC',
        circulating_supply: '18025852',
        max_supply: '21000000',
        name: 'Bitcoin',
        logo_url: 'https://s3.us-east-2.amazonaws.com/nomics-api/static/images/currencies/btc.svg',
        market_cap: '1243423423.05',
        rank: '1',
        high: '19480.90',
        high_timestamp: '2017-12-16',
      },
      {
        currecy: 'LTC',
        id: 'LTC',
        price: 75.55,
        price_date: '2019-11-02',
        symbol: 'LTC',
        circulating_supply: '18000000',
        max_supply: '20000000',
        name: 'Litecoin',
        logo_url: 'https://s3.us-east-2.amazonaws.com/nomics-api/static/images/currencies/ltc.svg',
        market_cap: '1003423423.05',
        rank: '3',
        high: '420.90',
        high_timestamp: '2017-12-10',
      },
    ];
    const mockTrackedCoins = ['BTC', 'LTC'];
    const mockState = {
      cryptos: mockCryptos,
      trackedCoins: mockTrackedCoins,
      error: 'There was an error',
    }
    const result = mapStateToProps(mockState);
    expect(result).toEqual({ cryptos: mockCryptos, trackedCoins: mockTrackedCoins });
  });
});

describe('mapDispatchToProps', () => {
  const mockDispatch = jest.fn();
  const mockTrackedCoins = ['BTC', 'LTC'];
  const mockCryptos = [
    {
      currecy: 'BTC',
      id: 'BTC',
      price: 9341.6987,
      price_date: '2019-11-02',
      symbol: 'BTC',
      circulating_supply: '18025852',
      max_supply: '21000000',
      name: 'Bitcoin',
      logo_url: 'https://s3.us-east-2.amazonaws.com/nomics-api/static/images/currencies/btc.svg',
      market_cap: '1243423423.05',
      rank: '1',
      high: '19480.90',
      high_timestamp: '2017-12-16',
    },
    {
      currecy: 'LTC',
      id: 'LTC',
      price: 75.55,
      price_date: '2019-11-02',
      symbol: 'LTC',
      circulating_supply: '18000000',
      max_supply: '20000000',
      name: 'Litecoin',
      logo_url: 'https://s3.us-east-2.amazonaws.com/nomics-api/static/images/currencies/ltc.svg',
      market_cap: '1003423423.05',
      rank: '3',
      high: '420.90',
      high_timestamp: '2017-12-10',
    },
  ];
  const trackedCoinsAction = actions.updateTrackedCoins(mockTrackedCoins);
  const cryptosAction = actions.updateCryptos(mockCryptos);
  const mappedDispatch = mapDispatchToProps(mockDispatch);
  mappedDispatch.updateTrackedCoins(mockTrackedCoins);
  expect(mockDispatch).toHaveBeenCalledWith(trackedCoinsAction);
  mappedDispatch.updateCryptos(mockCryptos);
  expect(mockDispatch).toHaveBeenCalledWith(cryptosAction);
});
