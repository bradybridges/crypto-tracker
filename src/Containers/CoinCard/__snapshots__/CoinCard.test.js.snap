// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CoinCard should match the snapshot 1`] = `
<section
  id="coin-card-section"
>
  <h1>
    Monero
  </h1>
  <img
    alt="coin logo"
    src="https://s3.us-east-2.amazonaws.com/nomics-api/static/images/currencies/xmr.svg"
  />
  <table>
    <tbody>
      <tr>
        <td>
          Price
        </td>
        <td>
          $
          1003.43
        </td>
      </tr>
      <tr>
        <td>
          Circulation
        </td>
        <td>
          123123413
        </td>
      </tr>
    </tbody>
  </table>
  <button
    className="tracking-btn"
    onClick={[Function]}
    type="button"
  >
    Track Coin
  </button>
</section>
`;
